jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpTransportOverHTTP2.java

```diff
@@ -350,119 +351,356 @@ public void abort(Throwable failure)
-            if (result)
+                LOG.debug("HTTP2 Response #{}/{} {} success",
+                    stream.getId(), Integer.toHexString(stream.getSession().hashCode()),
+                    commit ? "commit" : "flush");
+            callback.succeeded();
+        }
+
+        private void fail(Callback callback, boolean commit, Throwable failure)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("HTTP2 Response #{}/{} {} failure",
+                    stream.getId(), Integer.toHexString(stream.getSession().hashCode()),
+                    commit ? "commit" : "flush",
+                    failure);
+            if (callback != null)
                 callback.failed(failure);
-            return result;
         }
 
-        @Override
-        public InvocationType getInvocationType()
+        private void idleTimeout(Callback callback, boolean timeout, Throwable failure)
         {
-            Callback callback;
-            synchronized (this)
-            {
-                callback = this.callback;
-            }
-            return callback != null ? callback.getInvocationType() : Callback.super.getInvocationType();
+            if (LOG.isDebugEnabled())
+                LOG.debug("HTTP2 Response #{}/{} idle timeout {}",
+                    stream.getId(), Integer.toHexString(stream.getSession().hashCode()),
+                    timeout ? "expired" : "ignored",
+                    failure);
+            if (timeout)
+                callback.failed(failure);
         }
     }
 
+    /**
+     * <p>Send states for {@link TransportCallback}.</p>
+     *
+     * @see TransportCallback
+     */
     private enum State
     {
-        IDLE, WRITING, FAILED, TIMEOUT
+        /**
+         * <p>No send initiated or in progress.</p>
+         * <p>Next states could be:</p>
+         * <ul>
+         *   <li>{@link #PRE_SEND}, when {@link TransportCallback#send(Callback, boolean, Consumer)}
+         *   is called by the transport to initiate a send</li>
+         *   <li>{@link #FAILED}, when {@link TransportCallback#failed(Throwable)}
+         *   is called by an asynchronous failure</li>
+         * </ul>
+         */
+        IDLE,
+        /**
+         * <p>A send is initiated; the nested callback in {@link TransportCallback}
+         * cannot be notified while in this state.</p>
+         * <p>Next states could be:</p>
+         * <ul>
+         *   <li>{@link #SUCCEED}, when {@link TransportCallback#succeeded()}
+         *   is called synchronously because the send succeeded</li>
+         *   <li>{@link #FAIL}, when {@link TransportCallback#failed(Throwable)}
+         *   is called synchronously because the send failed</li>
+         *   <li>{@link #POST_SEND}, when {@link TransportCallback#postSend()}
+         *   is called before the send completes</li>
+         * </ul>
+         */
+        PRE_SEND,
```

Comment:
I don't like the `PRE_SEND`,  `POST_SEND` and `SUCCEED`names as they are confusing in that during `PRE_SEND` sending is happening and in `POST_SEND` the send is pending.  

How about: `IDLE` -> `SENDING` -> `PENDING` -> `SUCCEEDED` -> `FAILED`

I think `ING` and `ED` names line up better with our practises elsewhere.

Reply:
They are not correct.

SUCCEED and FAIL are commands, it's what `postSend()` should do, either succeed by calling `succeeded()` or fail by calling `failed(Throwable)`.

I'm ok with PRE_SEND -> SENDING and POST_SEND -> PENDING, but I need 2 more states for SUCCEED and FAIL _as commands_ about what to do next. 

FAILED is a terminal state (so it cannot be used for the command), but we don't have a SUCCEEDED terminal state (we go back to IDLE).